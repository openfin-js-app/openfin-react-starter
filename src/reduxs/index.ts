import {combineReducers} from 'redux';
import {IApplicationState} from './application/types';
import {IClientState} from './client/types';
import {IConfigState} from './config/types';

import application from './application/reducer';
import { defaultState as applicationDefaultState } from './application/reducer';
import client from './client/reducer';
import { defaultState as clientDefaultState } from './client/reducer';
import config from './config/reducer';
import { defaultState as configDefaultState } from './config/reducer';

export * from './application/actions';
export * from './application/types';
export * from './client/actions';
export * from './client/types';
export * from './config/actions';
export * from './config/context';
export * from './config/types';

export interface IRootState {
    application:IApplicationState,
    client:IClientState,
    config:IConfigState,
}

export const rootDefaultState = {
    application:applicationDefaultState,
    client:clientDefaultState,
    config:configDefaultState,
}

export default (parentWindowState?:IRootState)=>{
    if (parentWindowState){
        return combineReducers({
            application:application(parentWindowState.application),
            client:client(parentWindowState.client),
            config:config(parentWindowState.config),
        });

    }else{
        return combineReducers({
            application:application(),
            client:client(),
            config:config(),
        });
    }
}
