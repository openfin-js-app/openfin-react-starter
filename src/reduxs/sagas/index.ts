import { all, takeEvery, select } from 'redux-saga/effects';

import clientSaga from './client';

export function* handleLogAllActions(action) {
    const state = yield select();
    console.log(`${action.type}`,action,'state after',state);
}

export function* watchAndLogSaga(){
    yield takeEvery('*',handleLogAllActions)
}

export default function* rootSaga(){

    const sagas = [ clientSaga() ];

    if(process.env.REACT_APP_LOG_ACTION === 'true'){
        sagas.unshift(watchAndLogSaga());
    }

    yield all(sagas);

}