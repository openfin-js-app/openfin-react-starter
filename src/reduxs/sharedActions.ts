import {
    APPLICATION_SET_LOADING_MSG,
    APPLICATION_LAUNCH_NEW_WINDOW,
} from './application/actions';

import {
    CLIENT_SET_VALUE,
} from './client/actions';

const sharedActions:string[]=[
    APPLICATION_SET_LOADING_MSG,
    APPLICATION_LAUNCH_NEW_WINDOW,
    CLIENT_SET_VALUE,
];

export default sharedActions;