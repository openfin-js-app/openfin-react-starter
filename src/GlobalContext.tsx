import * as React from 'react';
import { connect } from 'react-redux'

import { I18Language, IConfigRuntimeState } from './reduxs'

import { ConfigContextProvider } from './reduxs/config/context';

interface IProps {
    config:IConfigRuntimeState,
    onToggleThemeField?:()=>void,
    onUpdateLangField?:(lang:I18Language)=>void,
}

class GlobalContext extends React.Component<IProps,{}>{

    render(){

        const {
            config,
            onToggleThemeField,
            onUpdateLangField,
        } = this.props;

        return (
            <ConfigContextProvider value={
                {
                    config,
                    actions:{
                        onToggleThemeField,
                        onUpdateLangField,
                    }
                }
            }>
                {this.props.children}
            </ConfigContextProvider>

        )
    }
}

export default GlobalContext;