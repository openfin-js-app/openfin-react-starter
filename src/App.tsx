import * as React from 'react';
import { Router, Route, Switch, Redirect } from 'react-router-dom';
import { connect } from 'react-redux'

import CssBaseline from '@material-ui/core/CssBaseline';
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';

import GlobalContext from './GlobalContext';

import { primaryColor, infoColor, dangerColor} from './assets/jss/openfin-starter-constant';

import indexRoutes from './routes/index';

import hist from './utils/history';

import {MuiTheme, I18Language, IRootState, IConfigRuntimeState, configUpdateOneField} from './reduxs'

interface IProps{
    loading:boolean,
    config:IConfigRuntimeState,
    theme:MuiTheme,
    actions:{
        handleUpdateThemeField:(value:MuiTheme)=>void,
        handleToggleThemeField:()=>void,
        handleUpdateLangField:(lang:I18Language)=>void,
    },
    [key:number]:any,
    [key:string]:any,
}

interface IState {
    [key:number]:any,
    [key:string]:any,
}


class App extends React.Component<IProps,IState>{

    createMuiTheme = ()=>{
        const { theme } = this.props;
        return createMuiTheme({
            typography:{
                useNextVariants:true,
                fontSize: 8,
                htmlFontSize: 10,
            },
            palette:{
                type:theme,
                primary:{
                    main:primaryColor,
                },
                secondary:{
                    main:infoColor,
                },
                error:{
                    main:dangerColor,
                },
            },
            overrides:{
                MuiButton:{
                    root:{
                        color:'white',
                    },
                    contained:{
                        padding:'0px 5px',
                        minWidth:'50px',
                        minHeight:'20px',
                        color:'white',
                    },
                    containedPrimary:{
                        color:'white',
                    },
                    containedSecondary:{
                        color:'white',
                    },
                    sizeSmall:{
                        padding:'0px 5px',
                        minWidth:'40px',
                        minHeight:'16px',
                    },
                    sizeLarge:{
                        padding:'0px 5px',
                        minWidth:'60px',
                        minHeight:'25px',
                    },
                },
                MuiChip:{
                    root:{
                        height:'20px',
                    },
                    avatar:{
                        width:'20px',
                        height:'20px',
                        fontSize:'0.5rem',
                    },
                },
                MuiFab:{
                    root:{
                        padding: '0px 0px',
                        width:'40px',
                        height:'40px',
                    }
                },
                MuiIconButton:{
                    root:{
                        padding: '0px 0px',
                    }
                },
                MuiListItem:{
                    gutters:{
                        paddingTop:'2px',
                        paddingBottom:'2px',
                    }
                },
                MuiSnackbarContent:{
                    root:{
                        color:'white',
                    }
                },
                MuiTabs:{
                    root:{
                        minHeight:'24px',
                    }
                },
                MuiTab:{
                    root:{
                        minHeight:'24px',
                    },
                    labelContainer:{
                        paddingTop:'0px',
                        paddingBottom:'0px',
                    }
                },
            }
        })
    }

    render(){

        const {
            config,
            actions:{
                handleToggleThemeField, handleUpdateLangField
            }
        } = this.props;

        return (
            <React.Fragment>
                <CssBaseline/>
                    <Router history={hist}>
                        <MuiThemeProvider theme={this.createMuiTheme()}>
                            <GlobalContext
                                config={config}
                                onToggleThemeField={handleToggleThemeField}
                                onUpdateLangField={handleUpdateLangField}
                            >
                                <Switch>
                                    {
                                        indexRoutes.map((prop:any,key)=>{
                                            if (prop.redirect)
                                                return <Redirect from={prop.path} to={prop.to} key={key}/>;
                                            return <Route path={prop.path} component={prop.component} key={key}/>;

                                        })
                                    }
                                </Switch>
                            </GlobalContext>
                        </MuiThemeProvider>
                    </Router>
            </React.Fragment>
        );
    }
}

export default connect(
    (state:IRootState)=>({
        loading:state.application.loading,
        config:state.config,
        theme:state.config.application.theme
    }),
    dispatch => ({
        actions:{
            handleUpdateThemeField: (value:MuiTheme)=>{
                dispatch(configUpdateOneField({
                    name:'application.theme',
                    value,
                }))
            },
            handleUpdateLangField: (value:I18Language)=>{
                dispatch(configUpdateOneField({
                    name:'application.language',
                    value,
                }))
            }
        }
    }),
    (stateProps,actionProps)=>({
        ...stateProps,
        actions:{
            ...actionProps.actions,
            handleToggleThemeField: ()=>{
                actionProps.actions.handleUpdateThemeField(
                    stateProps.theme === MuiTheme.DARK ?MuiTheme.LIGHT:MuiTheme.DARK
                )
            }
        }

    })
)(App);
