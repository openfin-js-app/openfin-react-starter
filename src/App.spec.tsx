import * as React from 'react';
import { Provider } from 'react-redux';

import { createShallow, createMount, createRender } from '@material-ui/core/test-utils';
import configurestore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';

import GlobalContext from './GlobalContext';

import {I18Language, rootDefaultState} from './reduxs';

import App from './App';

const mockStore = configurestore();
const store = mockStore(rootDefaultState);

describe('App entry',()=>{

    let shallow;
    let mount;
    let render;

    beforeAll(() => {
        mount = createMount();
        render = createRender();
    });

    afterAll(() => {
        mount.cleanUp();
    });


    beforeEach(() => {
        shallow = createShallow();
    });

    it('it renders correctly by default',()=>{
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/']}>
                    <App location={{pathname:'/'}}
                    />
                </MemoryRouter>
            </Provider>
        )

        const gCtx = wrapper.find(GlobalContext);

        expect(gCtx).toHaveLength(1);
        gCtx.at(0).props().onToggleThemeField();
        gCtx.at(0).props().onUpdateLangField(I18Language.en_US);

        expect(store.getActions()).toMatchSnapshot();

        expect(wrapper).toBeTruthy();
    })

});