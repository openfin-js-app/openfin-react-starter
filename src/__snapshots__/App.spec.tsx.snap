// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App entry it renders correctly by default 1`] = `
Array [
  Object {
    "payload": Object {
      "name": "application.theme",
      "value": "light",
    },
    "type": "CONFIG_UPDATE_ONE_FIELD",
  },
  Object {
    "payload": Object {
      "name": "application.language",
      "value": "en-US",
    },
    "type": "CONFIG_UPDATE_ONE_FIELD",
  },
]
`;
